pkg/results/error.go

```diff
@@ -0,0 +1,83 @@
+package results
+
+import (
+	"fmt"
+)
+
+// Error holds a message and a child, allowing for an error
+// The common use-case here will be to wrap errors from callsites:
+//
+// if err := doSomething(data); err != nil {
+//     return results.ForReason(results.ReasonFoo).WithError(err).Errorf("could not do something for data: %v", data)
+// }
+type Error struct {
+	reason  Reason
+	message string
+	wrapped error
+}
+
+// Error makes an Error an error
+func (e Error) Error() string {
+	return e.message
+}
+
+// FullReason provides the chain of error reasons, divided by colons
+func (e Error) FullReason() string {
+	reasonedError, ok := e.wrapped.(*Error)
+	if !ok {
+		return string(e.reason)
+	}
+	return fmt.Sprintf("%s:%s", e.reason, reasonedError.FullReason())
+}
+
+// FullReason attempts to get the full reason from an error, or uses
+// unknown when it's not something we can do
+func FullReason(err error) string {
+	reasonedError, ok := err.(*Error)
+	if !ok {
+		return string(ReasonUnknown)
+	}
+	return reasonedError.FullReason()
+}
+
+// ForReason is a constructor for an Error from a Reason
+func ForReason(reason Reason) *Error {
+	if reason == "" {
+		// we don't want to publish metrics with an empty label, so
+		// we enforce that there's some default (if useless) value
+		reason = ReasonUnknown
+	}
+	return &Error{
+		reason: reason,
+	}
+}
+
+// WithError is a builder that adds a child to the Error
+func (e *Error) WithError(err error) *Error {
+	e.wrapped = err
+	return e
+}
+
+// Errorf is a builder that adds in the main error to an Error
+func (e *Error) Errorf(format string, args ...interface{}) error {
+	e.message = fmt.Sprintf(format, args...)
```

Comment:
Please call it `SetMessagef`, its unclear from the godocs what this does. Also, why does this return an `error` and not an `*Error`?

Reply:
It's the producer, so you end up with an `error` as you expect

Reply:
Great, the same argument can be made for every single other public function in this package. Please make all of them either return an `error`, `*Error`, or `Error`.

You also completely ignored the first part of my message, without reading the implementation its completely unclear what this does